modules/apps/login/login-authentication-opensso-web/src/main/java/com/liferay/login/authentication/opensso/web/internal/portlet/action/OpenSSOMVCRenderCommand.java

```
@@ -0,0 +1,187 @@
+import com.liferay.portal.security.sso.opensso.constants.OpenSSOConstants;
+
+import java.io.IOException;
+
+import javax.portlet.PortletException;
+import javax.portlet.RenderRequest;
+import javax.portlet.RenderResponse;
+
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletContext;
+import javax.servlet.http.HttpServletRequest;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Stian Sigvartsen
+ */
+@Component(
+	immediate = true,
+	property = {
+		"javax.portlet.name=" + PortletKeys.FAST_LOGIN,
+		"javax.portlet.name=" + PortletKeys.LOGIN, "mvc.command.name=/",
+		"mvc.command.name=/login/login", "service.ranking:Integer=1"
+	},
+	service = MVCRenderCommand.class
+)
+public class OpenSSOMVCRenderCommand implements MVCRenderCommand {
+
+	@Override
+	public String render(
+			RenderRequest renderRequest, RenderResponse renderResponse)
+		throws PortletException {
+
+		HttpServletRequest httpServletRequest = _portal.getHttpServletRequest(
+			renderRequest);
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		long companyId = themeDisplay.getCompanyId();
+
+		OpenSSOConfiguration openSSOConfiguration;
+
+		try {
+			openSSOConfiguration = getOpenSSOConfiguration(companyId);
+
+			if (!openSSOConfiguration.enabled() ||
+				Validator.isNull(openSSOConfiguration.loginURL()) ||
+				Validator.isNull(openSSOConfiguration.logoutURL()) |
+				Validator.isNull(openSSOConfiguration.serviceURL())) {
+
+				return "/login.jsp";
+			}
+		}
+		catch (Exception e) {
+			_log.error(e, e);
+
+			return "/login.jsp";
+		}
+
+		HttpServletRequest originalHttpServletRequest =
+			_portal.getOriginalServletRequest(httpServletRequest);
+
+		String error = (String)originalHttpServletRequest.getAttribute(
+			"open.sso.error");
+
+		if (Validator.isNotNull(error)) {
```

This should be the first check to fail fast.

---

I don't think we can optimize this because the command also handles no errors as per earlier comment